rfcs/0018-glides-trainsheet-events.md

```diff
@@ -0,0 +1,302 @@
+- Only 1 trainset (3800 and 3850) is available
+- In order to maintain headways, Inspector Alice will split the trainset
+- Operator A will take 3800 and run the 9:55am
+- Operator B will take 3850 and run the 10:00am
+
+```json
+{
+  "specversion": "1.0",
+  "type": "com.mbta.ctd.glides.operator-signed-in.v1",
+  "source": "glides.mbta.com",
+  "id": "19fdb184-7dd6-4664-8472-04bd6177ec44",
+  "time": "2023-01-20T09:45:10-05:00",
+  "data": {
+    "actor": {
+      "username": "ainspector",
+      "badgeNumber": "123"
+    },
+    "operator": {
+      "badgeNumber": "456"
+    },
+    "signInAt": "2023-01-20T09:45:00-05:00"
+  }
+}
+# further specversion, source, time, and ID fields elided from the top level
+# actor elided from underneath "data"
+{
+  "type": "com.mbta.ctd.glides.operator-signed-in.v1",
+  "data": {
+    "operator": {
+      "badgeNumber": "567"
+    },
+    "signInAt": "2023-01-20T09:45:00-05:00"
+  }
+}
+{
+  "type": "com.mbta.ctd.glides.trip-dropped.v1",
+  "data": {
+    "tripKey": {
+      "serviceDate": "2022-01-20",
+      "stop": "place-lake",
+      "scheduledTime": "10:00:00"
+    },
+    "reason": "ran as single"
+  }
+}
+{
+  "type": "com.mbta.ctd.glides.trip-updated.v1",
+  "data": {
+    "tripKey": {
+      "serviceDate": "2022-01-20",
+      "stop": "place-lake",
+      "scheduledTime": "09:55:00"
+    },
+    "operators": [
+      {"badgeNumber": "456"}
+    ],
+    "consist": [
+      {"carNumber": "3800"}
+    ]
+  }
+}
+{
+  "type": "com.mbta.ctd.glides.trip-added.v1",
+  "time": "2023-01-20T09:50:15-05:00",
+  "data": {
+    "actor": {
+      "username": "ainspector",
+      "badgeNumber": "123"
+    },
+    "tripKey": {
+      "service-date": "2022-01-20",
+      "id": "ADDED-1"
+    },
+    "stop": "place-lake",
+    "time": "10:00:00",
+    "tripType": "round",
+    "consist": [
+      {"carNumber": "3850"}
+    ],
+    "operators": [
+      {"badgeNumber": "567"}
+    ]
+  }
+}
+```
+
+In this instance, the inspector dropped the 10:00am trip, and created a new trip with the same departure time. Ideally, RTR will treat this the same as if the 10:00am were updated to be a 1-car trip, but inspectors do not always enter the data in that fashion.
+
+# Drawbacks
+- This requires Glides to include new functionality to write their messages into a Kinesis stream. However, they already use `ex_aws` so additional messages should not be hard to add.
+ 
+# Rationale and alternatives
+- RTR already consumes a Kinesis stream of CloudEvents from OCS, 
+- so it's a limited additional lift to consume a new Kinesis stream
+- Kinesis Firehose allows writing the stream of messages to S3 for LAMP to consume, without needing a separate integration
+
+## Alternative: split update types into different events
+Examples:
+- set departure time
+- set/unset operator
+- set/unset consist
+
+The current trainsheet interface does not distinguish between these cases, so it would need to be added to Glides.
```

Comment:
I'm confused by this sentence. There is a distinction between changing a time and changing an operator, because those are different fields.

Reply:
Glides internally has a single "update all the information about a trip" event. It wouldn't be that hard to break that into a batch of smaller events for the event stream.

One potential benefit: If we add new data (for example, marking trains as nonrevenue), in the single big "updated-trip" event, it requires adding a field to the existing event. If we broke updates into many smaller events, then it would be a new type of event, which could maybe make backwards compatibility easier.

(I still prefer one big event)

Reply:
They are different fields, but as Sky noted, the only interface currently is to save all changes at once.